File: Sources/CodexRadarCore/QuotaHistory.swift
import Darwin
import Foundation

public enum QuotaHistoryRange: String, CaseIterable, Codable, Sendable {
    case hours24
    case days7
    case days30

    public var duration: TimeInterval {
        switch self {
        case .hours24:
            return 24 * 60 * 60
        case .days7:
            return 7 * 24 * 60 * 60
        case .days30:
            return 30 * 24 * 60 * 60
        }
    }

    public var displayBucketDuration: TimeInterval {
        switch self {
        case .hours24:
            return 15 * 60
        case .days7:
            return 60 * 60
        case .days30:
            return 4 * 60 * 60
        }
    }

    public var continuityGap: TimeInterval {
        switch self {
        case .hours24:
            return 20 * 60
        case .days7:
            return 2 * 60 * 60
        case .days30:
            return 8 * 60 * 60
        }
    }
}

public struct QuotaHistorySample: Codable, Equatable, Identifiable, Sendable {
    public let timestamp: Date
    public let remainingPercent: Double
    public let resetsAt: Date?

    public var id: Date {
        timestamp
    }

    public init(
        timestamp: Date,
        remainingPercent: Double,
        resetsAt: Date?
    ) {
        self.timestamp = timestamp
        self.remainingPercent = remainingPercent
        self.resetsAt = resetsAt
    }
}

public struct QuotaHistoryResetEvent: Equatable, Identifiable, Sendable {
    public let timestamp: Date
    public let previousRemainingPercent: Double
    public let remainingPercent: Double

    public var id: Date {
        timestamp
    }

    public var increase: Double {
        remainingPercent - previousRemainingPercent
    }
}

public struct QuotaHistorySummary: Equatable, Sendable {
    public let observedConsumption: Double
    public let resetCount: Int
    public let sampleCount: Int

    public init(
        observedConsumption: Double,
        resetCount: Int,
        sampleCount: Int
    ) {
        self.observedConsumption = observedConsumption
        self.resetCount = resetCount
        self.sampleCount = sampleCount
    }
}

public struct QuotaHistoryTimeline: Equatable, Sendable {
    public static let archiveVersion = 1
    public static let retentionDuration: TimeInterval = 31 * 24 * 60 * 60

    public private(set) var samples: [QuotaHistorySample]

    public init(samples: [QuotaHistorySample] = []) {
        var byTimestamp: [Date: QuotaHistorySample] = [:]
        for sample in samples where Self.isValid(sample) {
            byTimestamp[sample.timestamp] = sample
        }
        self.samples = byTimestamp.values.sorted {
            $0.timestamp < $1.timestamp
        }
    }

    @discardableResult
    public mutating func record(
        _ sample: QuotaHistorySample,
        minimumInterval: TimeInterval = 5 * 60,
        minimumChange: Double = 0.25,
        endingAt now: Date
    ) -> Bool {
        guard Self.isValid(sample),
              sample.timestamp <= now.addingTimeInterval(5 * 60) else {
            return false
        }

        prune(endingAt: now)
        guard let previous = samples.last else {
            samples.append(sample)
            return true
        }
        guard sample.timestamp > previous.timestamp else {
            return false
        }

        let interval = sample.timestamp.timeIntervalSince(previous.timestamp)
        let change = abs(
            sample.remainingPercent - previous.remainingPercent
        )
        let resetBoundaryChanged = sample.resetsAt != previous.resetsAt
        guard interval >= minimumInterval
                || change >= minimumChange
                || resetBoundaryChanged else {
            return false
        }

        samples.append(sample)
        prune(endingAt: now)
        return true
    }

    public mutating func prune(endingAt now: Date) {
        let cutoff = now.addingTimeInterval(-Self.retentionDuration)
        samples.removeAll { $0.timestamp < cutoff }
    }

    public func samples(
        in range: QuotaHistoryRange,
        endingAt now: Date
    ) -> [QuotaHistorySample] {
        let start = now.addingTimeInterval(-range.duration)
        return samples.filter {
            $0.timestamp >= start && $0.timestamp <= now
        }
    }

    public func displaySamples(
        in range: QuotaHistoryRange,
        endingAt now: Date
    ) -> [QuotaHistorySample] {
        let visible = samples(in: range, endingAt: now)
        guard visible.count > 2 else {
            return visible
        }

        let start = now.addingTimeInterval(-range.duration)
        let resetTimes = Set(
            resetEvents(in: range, endingAt: now).map(\.timestamp)
        )
        var selected = Set<Date>()
        var buckets: [Int: [QuotaHistorySample]] = [:]

        for sample in visible {
            let offset = sample.timestamp.timeIntervalSince(start)
            let bucket = Int(floor(offset / range.displayBucketDuration))
            buckets[bucket, default: []].append(sample)
        }

        for bucket in buckets.values {
            guard let first = bucket.first, let last = bucket.last else {
                continue
            }
            selected.insert(first.timestamp)
            selected.insert(last.timestamp)

            if let minimum = bucket.min(
                by: { $0.remainingPercent < $1.remainingPercent }
            ) {
                selected.insert(minimum.timestamp)
            }
            if let maximum = bucket.max(
                by: { $0.remainingPercent < $1.remainingPercent }
            ) {
                selected.insert(maximum.timestamp)
            }
            for sample in bucket where resetTimes.contains(sample.timestamp) {
                selected.insert(sample.timestamp)
            }
        }

        return visible.filter { selected.contains($0.timestamp) }
    }

    public func resetEvents(
        in range: QuotaHistoryRange,
        endingAt now: Date
    ) -> [QuotaHistoryResetEvent] {
        let visible = samples(in: range, endingAt: now)
        guard visible.count >= 2 else {
            return []
        }

        return zip(visible, visible.dropFirst()).compactMap {
            previous, current in
            guard Self.isObservedReset(
                previous: previous,
                current: current
            ) else {
                return nil
            }
            return QuotaHistoryResetEvent(
                timestamp: current.timestamp,
                previousRemainingPercent: previous.remainingPercent,
                remainingPercent: current.remainingPercent
            )
        }
    }

    public func summary(
        in range: QuotaHistoryRange,
        endingAt now: Date
    ) -> QuotaHistorySummary {
        let visible = samples(in: range, endingAt: now)
        let consumption = zip(visible, visible.dropFirst()).reduce(0.0) {
            partial, pair in
            partial + max(
                0,
                pair.0.remainingPercent - pair.1.remainingPercent
            )
        }
        return QuotaHistorySummary(
            observedConsumption: consumption,
            resetCount: resetEvents(in: range, endingAt: now).count,
            sampleCount: visible.count
        )
    }

    public func nearestSample(
        to date: Date,
        in range: QuotaHistoryRange,
        endingAt now: Date
    ) -> QuotaHistorySample? {
        let visible = samples(in: range, endingAt: now)
        guard !visible.isEmpty else {
            return nil
        }

        var lower = 0
        var upper = visible.count
        while lower < upper {
            let middle = (lower + upper) / 2
            if visible[middle].timestamp < date {
                lower = middle + 1
            } else {
                upper = middle
            }
        }

        if lower == 0 {
            return visible[0]
        }
        if lower == visible.count {
            return visible[visible.count - 1]
        }
        let before = visible[lower - 1]
        let after = visible[lower]
        return date.timeIntervalSince(before.timestamp)
            <= after.timestamp.timeIntervalSince(date)
            ? before
            : after
    }

    public func previousSample(
        before sample: QuotaHistorySample
    ) -> QuotaHistorySample? {
        guard let index = samples.firstIndex(
            where: { $0.timestamp == sample.timestamp }
        ), index > 0 else {
            return nil
        }
        return samples[index - 1]
    }

    public func isResetSample(_ sample: QuotaHistorySample) -> Bool {
        guard let previous = previousSample(before: sample) else {
            return false
        }
        return Self.isObservedReset(previous: previous, current: sample)
    }

    static func isValid(_ sample: QuotaHistorySample) -> Bool {
        sample.timestamp.timeIntervalSinceReferenceDate.isFinite
            && sample.remainingPercent.isFinite
            && (0...100).contains(sample.remainingPercent)
            && (
                sample.resetsAt == nil
                    || sample.resetsAt?.timeIntervalSinceReferenceDate
                    .isFinite == true
            )
    }

    static func isObservedReset(
        previous: QuotaHistorySample,
        current: QuotaHistorySample
    ) -> Bool {
        let increase =
            current.remainingPercent - previous.remainingPercent
        guard increase > 0 else {
            return false
        }

        if let previousReset = previous.resetsAt,
           let currentReset = current.resetsAt,
           currentReset.timeIntervalSince(previousReset) >= 30 * 60,
           increase >= 5 {
            return true
        }
        if current.remainingPercent >= 80, increase >= 20 {
            return true
        }
        return increase >= 50
    }
}

public struct QuotaHistoryRecordResult: Equatable, Sendable {
    public let timeline: QuotaHistoryTimeline
    public let didChange: Bool

    public init(timeline: QuotaHistoryTimeline, didChange: Bool) {
        self.timeline = timeline
        self.didChange = didChange
    }
}

public struct QuotaHistoryStore: Sendable {
    public enum LoadResult: Equatable, Sendable {
        case absent
        case loaded(QuotaHistoryTimeline)
        case corrupt
        case unavailable
    }

    public enum StorageError: LocalizedError {
        case cannotPrepareDirectory
        case cannotLock
        case corruptArchive
        case cannotEncode
        case cannotWrite
        case cannotSync
        case cannotReplace

        public var errorDescription: String? {
            switch self {
            case .cannotPrepareDirectory:
                return "Quota history directory is unavailable"
            case .cannotLock:
                return "Quota history is being updated by another process"
            case .corruptArchive:
                return "Quota history cannot be decoded safely"
            case .cannotEncode:
                return "Quota history cannot be encoded"
            case .cannotWrite:
                return "Quota history cannot be written"
            case .cannotSync:
                return "Quota history cannot be synchronized"
            case .cannotReplace:
                return "Quota history cannot be committed"
            }
        }
    }

    public let url: URL
    private let lockURL: URL

    public init(url: URL = Self.defaultURL) {
        self.url = url
        self.lockURL = url
            .deletingPathExtension()
            .appendingPathExtension("lock")
    }

    public static var defaultURL: URL {
        let base = FileManager.default.urls(
            for: .applicationSupportDirectory,
            in: .userDomainMask
        ).first ?? FileManager.default.temporaryDirectory
        return base
            .appendingPathComponent(
                AppConstants.bundleIdentifier,
                isDirectory: true
            )
            .appendingPathComponent("weekly-quota-history-v1.json")
    }

    public func load(endingAt now: Date = Date()) -> LoadResult {
        do {
            return try withLock(operation: LOCK_SH) {
                try loadUnlocked(endingAt: now)
            }
        } catch {
            return .unavailable
        }
    }

    public func record(
        _ sample: QuotaHistorySample,
        endingAt now: Date = Date()
    ) throws -> QuotaHistoryRecordResult {
        try withLock(operation: LOCK_EX) {
            var timeline: QuotaHistoryTimeline
            switch try loadUnlocked(endingAt: now) {
            case .absent:
                timeline = QuotaHistoryTimeline()
            case .loaded(let loaded):
                timeline = loaded
            case .corrupt:
                throw StorageError.corruptArchive
            case .unavailable:
                throw StorageError.cannotWrite
            }

            let changed = timeline.record(sample, endingAt: now)
            guard changed else {
                return QuotaHistoryRecordResult(
                    timeline: timeline,
                    didChange: false
                )
            }
            try saveUnlocked(timeline)
            return QuotaHistoryRecordResult(
                timeline: timeline,
                didChange: true
            )
        }
    }

    private struct Archive: Codable {
        let version: Int
        let samples: [QuotaHistorySample]
    }

    private func loadUnlocked(endingAt now: Date) throws -> LoadResult {
        guard FileManager.default.fileExists(atPath: url.path) else {
            return .absent
        }
        guard let data = try? Data(contentsOf: url),
              let archive = try? JSONDecoder().decode(
                  Archive.self,
                  from: data
              ),
              archive.version == QuotaHistoryTimeline.archiveVersion,
              archive.samples.count <= 100_000,
              zip(
                  archive.samples,
                  archive.samples.dropFirst()
              ).allSatisfy({ $0.timestamp < $1.timestamp }),
              archive.samples.allSatisfy(QuotaHistoryTimeline.isValid) else {
            return .corrupt
        }

        var timeline = QuotaHistoryTimeline(samples: archive.samples)
        timeline.prune(endingAt: now)
        return .loaded(timeline)
    }

    private func saveUnlocked(_ timeline: QuotaHistoryTimeline) throws {
        let archive = Archive(
            version: QuotaHistoryTimeline.archiveVersion,
            samples: timeline.samples
        )
        guard let data = try? JSONEncoder().encode(archive) else {
            throw StorageError.cannotEncode
        }
        try Self.atomicWrite(data, to: url)
    }

    private func withLock<T>(
        operation: Int32,
        _ body: () throws -> T
    ) throws -> T {
        let directory = url.deletingLastPathComponent()
        do {
            try FileManager.default.createDirectory(
                at: directory,
                withIntermediateDirectories: true,
                attributes: [.posixPermissions: 0o700]
            )
            try FileManager.default.setAttributes(
                [.posixPermissions: 0o700],
                ofItemAtPath: directory.path
            )
        } catch {
            throw StorageError.cannotPrepareDirectory
        }

        let descriptor = lockURL.path.withCString {
            Darwin.open(
                $0,
                O_CREAT | O_RDWR | O_CLOEXEC,
                S_IRUSR | S_IWUSR
            )
        }
        guard descriptor >= 0,
              Darwin.fchmod(
                  descriptor,
                  S_IRUSR | S_IWUSR
              ) == 0 else {
            if descriptor >= 0 {
                Darwin.close(descriptor)
            }
            throw StorageError.cannotLock
        }
        while flock(descriptor, operation) != 0 {
            if errno == EINTR {
                continue
            }
            Darwin.close(descriptor)
            throw StorageError.cannotLock
        }
        defer {
            flock(descriptor, LOCK_UN)
            Darwin.close(descriptor)
        }
        return try body()
    }

    private static func atomicWrite(_ data: Data, to url: URL) throws {
        let directory = url.deletingLastPathComponent()
        let temporaryURL = directory.appendingPathComponent(
            ".\(url.lastPathComponent).\(UUID().uuidString).tmp"
        )
        let descriptor = temporaryURL.path.withCString {
            Darwin.open(
                $0,
                O_CREAT | O_EXCL | O_WRONLY | O_CLOEXEC,
                S_IRUSR | S_IWUSR
            )
        }
        guard descriptor >= 0 else {
            throw StorageError.cannotWrite
        }

        var descriptorIsOpen = true
        var shouldRemoveTemporary = true
        defer {
            if descriptorIsOpen {
                Darwin.close(descriptor)
            }
            if shouldRemoveTemporary {
                try? FileManager.default.removeItem(at: temporaryURL)
            }
        }

        let wroteAllBytes = data.withUnsafeBytes { buffer -> Bool in
            guard let baseAddress = buffer.baseAddress else {
                return data.isEmpty
            }
            var written = 0
            while written < buffer.count {
                let result = Darwin.write(
                    descriptor,
                    baseAddress.advanced(by: written),
                    buffer.count - written
                )
                if result < 0, errno == EINTR {
                    continue
                }
                guard result > 0 else {
                    return false
                }
                written += result
            }
            return true
        }
        guard wroteAllBytes else {
            throw StorageError.cannotWrite
        }
        guard Darwin.fsync(descriptor) == 0 else {
            throw StorageError.cannotSync
        }
        let closeResult = Darwin.close(descriptor)
        descriptorIsOpen = false
        guard closeResult == 0 else {
            throw StorageError.cannotSync
        }
        let renameResult = temporaryURL.path.withCString { temporaryPath in
            url.path.withCString { destinationPath in
                Darwin.rename(temporaryPath, destinationPath)
            }
        }
        guard renameResult == 0 else {
            throw StorageError.cannotReplace
        }
        shouldRemoveTemporary = false

        let directoryDescriptor = directory.path.withCString {
            Darwin.open($0, O_RDONLY | O_CLOEXEC)
        }
        guard directoryDescriptor >= 0 else {
            throw StorageError.cannotSync
        }
        defer {
            Darwin.close(directoryDescriptor)
        }
        guard Darwin.fsync(directoryDescriptor) == 0 else {
            throw StorageError.cannotSync
        }
    }
}

File: windows/CodexRadar.Windows/QuotaHistory.cs
using System.Text.Json;

namespace CodexRadar.Windows;

internal static class QuotaHistoryObservationPolicy
{
    public static QuotaHistorySample? CreateSample(
        DashboardSnapshot snapshot,
        DateTimeOffset? previousObservation,
        DashboardPreview preview)
    {
        if (preview != DashboardPreview.Live
            || snapshot.QuotaObservedAt is not { } observedAt
            || observedAt == previousObservation
            || snapshot.WeeklyUsedPercent is not { } weeklyUsed)
            return null;

        return new QuotaHistorySample(
            observedAt,
            Math.Clamp(100 - weeklyUsed, 0, 100),
            snapshot.WeeklyResetsAt);
    }
}

internal static class QuotaHistoryRangeExtensions
{
    public static TimeSpan Duration(this QuotaHistoryRange range) => range switch
    {
        QuotaHistoryRange.Hours24 => TimeSpan.FromHours(24),
        QuotaHistoryRange.Days7 => TimeSpan.FromDays(7),
        _ => TimeSpan.FromDays(30)
    };

    public static TimeSpan DisplayBucketDuration(this QuotaHistoryRange range) => range switch
    {
        QuotaHistoryRange.Hours24 => TimeSpan.FromMinutes(15),
        QuotaHistoryRange.Days7 => TimeSpan.FromHours(1),
        _ => TimeSpan.FromHours(4)
    };

    public static TimeSpan ContinuityGap(this QuotaHistoryRange range) => range switch
    {
        QuotaHistoryRange.Hours24 => TimeSpan.FromMinutes(20),
        QuotaHistoryRange.Days7 => TimeSpan.FromHours(2),
        _ => TimeSpan.FromHours(8)
    };
}

internal sealed record QuotaHistorySample(
    DateTimeOffset Timestamp,
    double RemainingPercent,
    DateTimeOffset? ResetsAt);

internal sealed record QuotaHistoryResetEvent(
    DateTimeOffset Timestamp,
    double PreviousRemainingPercent,
    double RemainingPercent)
{
    public double Increase => RemainingPercent - PreviousRemainingPercent;
}

internal sealed record QuotaHistorySummary(
    double ObservedConsumption,
    int ResetCount,
    int SampleCount);

internal sealed class QuotaHistoryTimeline
{
    public const int ArchiveVersion = 1;
    public static readonly TimeSpan RetentionDuration = TimeSpan.FromDays(31);

    private readonly List<QuotaHistorySample> _samples;
    public IReadOnlyList<QuotaHistorySample> Samples => _samples;

    public QuotaHistoryTimeline(
        IEnumerable<QuotaHistorySample>? samples = null)
    {
        _samples = (samples ?? [])
            .Where(IsValid)
            .GroupBy(sample => sample.Timestamp)
            .Select(group => group.Last())
            .OrderBy(sample => sample.Timestamp)
            .ToList();
    }

    public bool Record(
        QuotaHistorySample sample,
        DateTimeOffset endingAt,
        TimeSpan? minimumInterval = null,
        double minimumChange = 0.25)
    {
        if (!IsValid(sample)
            || sample.Timestamp > endingAt.AddMinutes(5))
            return false;

        Prune(endingAt);
        if (_samples.Count == 0)
        {
            _samples.Add(sample);
            return true;
        }

        var previous = _samples[^1];
        if (sample.Timestamp <= previous.Timestamp)
            return false;

        var interval = sample.Timestamp - previous.Timestamp;
        var change = Math.Abs(
            sample.RemainingPercent
            - previous.RemainingPercent);
        var resetBoundaryChanged =
            sample.ResetsAt != previous.ResetsAt;
        if (interval < (minimumInterval ?? TimeSpan.FromMinutes(5))
            && change < minimumChange
            && !resetBoundaryChanged)
            return false;

        _samples.Add(sample);
        Prune(endingAt);
        return true;
    }

    public void Prune(DateTimeOffset endingAt)
    {
        var cutoff = endingAt - RetentionDuration;
        _samples.RemoveAll(sample => sample.Timestamp < cutoff);
    }

    public IReadOnlyList<QuotaHistorySample> SamplesIn(
        QuotaHistoryRange range,
        DateTimeOffset endingAt)
    {
        var start = endingAt - range.Duration();
        return _samples
            .Where(sample =>
                sample.Timestamp >= start
                && sample.Timestamp <= endingAt)
            .ToArray();
    }

    public IReadOnlyList<QuotaHistorySample> DisplaySamples(
        QuotaHistoryRange range,
        DateTimeOffset endingAt)
    {
        var visible = SamplesIn(range, endingAt);
        if (visible.Count <= 2) return visible;

        var start = endingAt - range.Duration();
        var resetTimes = ResetEvents(range, endingAt)
            .Select(item => item.Timestamp)
            .ToHashSet();
        var selected = new HashSet<DateTimeOffset>();
        var buckets = visible.GroupBy(sample =>
            (long)Math.Floor(
                (sample.Timestamp - start).TotalSeconds
                / range.DisplayBucketDuration().TotalSeconds));

        foreach (var bucket in buckets)
        {
            var items = bucket.ToArray();
            selected.Add(items[0].Timestamp);
            selected.Add(items[^1].Timestamp);
            selected.Add(items.MinBy(
                sample => sample.RemainingPercent)!.Timestamp);
            selected.Add(items.MaxBy(
                sample => sample.RemainingPercent)!.Timestamp);
            foreach (var sample in items
                         .Where(sample =>
                             resetTimes.Contains(sample.Timestamp)))
                selected.Add(sample.Timestamp);
        }

        return visible
            .Where(sample => selected.Contains(sample.Timestamp))
            .ToArray();
    }

    public IReadOnlyList<QuotaHistoryResetEvent> ResetEvents(
        QuotaHistoryRange range,
        DateTimeOffset endingAt)
    {
        var visible = SamplesIn(range, endingAt);
        if (visible.Count < 2) return [];

        var events = new List<QuotaHistoryResetEvent>();
        for (var index = 1; index < visible.Count; index++)
        {
            var previous = visible[index - 1];
            var current = visible[index];
            if (!IsObservedReset(previous, current)) continue;
            events.Add(new QuotaHistoryResetEvent(
                current.Timestamp,
                previous.RemainingPercent,
                current.RemainingPercent));
        }
        return events;
    }

    public QuotaHistorySummary Summary(
        QuotaHistoryRange range,
        DateTimeOffset endingAt)
    {
        var visible = SamplesIn(range, endingAt);
        var consumption = 0d;
        for (var index = 1; index < visible.Count; index++)
            consumption += Math.Max(
                0,
                visible[index - 1].RemainingPercent
                - visible[index].RemainingPercent);
        return new QuotaHistorySummary(
            consumption,
            ResetEvents(range, endingAt).Count,
            visible.Count);
    }

    public QuotaHistorySample? NearestSample(
        DateTimeOffset timestamp,
        QuotaHistoryRange range,
        DateTimeOffset endingAt)
    {
        var visible = SamplesIn(range, endingAt);
        if (visible.Count == 0) return null;

        var lower = 0;
        var upper = visible.Count;
        while (lower < upper)
        {
            var middle = (lower + upper) / 2;
            if (visible[middle].Timestamp < timestamp)
                lower = middle + 1;
            else
                upper = middle;
        }

        if (lower == 0) return visible[0];
        if (lower == visible.Count) return visible[^1];
        var before = visible[lower - 1];
        var after = visible[lower];
        return timestamp - before.Timestamp
               <= after.Timestamp - timestamp
            ? before
            : after;
    }

    public QuotaHistorySample? PreviousSample(
        QuotaHistorySample sample)
    {
        var index = _samples.FindIndex(
            item => item.Timestamp == sample.Timestamp);
        return index > 0 ? _samples[index - 1] : null;
    }

    public bool IsResetSample(QuotaHistorySample sample) =>
        PreviousSample(sample) is { } previous
        && IsObservedReset(previous, sample);

    public static QuotaHistoryTimeline CreatePreview(
        DateTimeOffset endingAt)
    {
        var samples = new List<QuotaHistorySample>();
        var start = endingAt - RetentionDuration;
        var resetAt = start.AddDays(7);
        var remaining = 96d;
        var index = 0;
        for (var time = start;
             time <= endingAt;
             time = time.AddHours(2), index++)
        {
            if (index is >= 210 and <= 218)
                continue;
            if (time >= resetAt)
            {
                remaining = 98;
                resetAt = resetAt.AddDays(7);
            }
            else
            {
                remaining = Math.Max(
                    8,
                    remaining - (index % 4 == 0 ? 1.4 : 0.35));
            }
            samples.Add(new QuotaHistorySample(
                time,
                remaining,
                resetAt));
        }
        return new QuotaHistoryTimeline(samples);
    }

    internal static bool IsValid(QuotaHistorySample sample) =>
        double.IsFinite(sample.RemainingPercent)
        && sample.RemainingPercent is >= 0 and <= 100;

    internal static bool IsObservedReset(
        QuotaHistorySample previous,
        QuotaHistorySample current)
    {
        var increase =
            current.RemainingPercent - previous.RemainingPercent;
        if (increase <= 0) return false;

        if (previous.ResetsAt is { } previousReset
            && current.ResetsAt is { } currentReset
            && currentReset - previousReset >= TimeSpan.FromMinutes(30)
            && increase >= 5)
            return true;
        if (current.RemainingPercent >= 80 && increase >= 20)
            return true;
        return increase >= 50;
    }
}

internal enum QuotaHistoryLoadStatus
{
    Absent,
    Loaded,
    Corrupt,
    Unavailable
}

internal sealed record QuotaHistoryLoadResult(
    QuotaHistoryLoadStatus Status,
    QuotaHistoryTimeline Timeline);

internal sealed record QuotaHistoryRecordResult(
    QuotaHistoryTimeline Timeline,
    bool DidChange);

internal sealed class QuotaHistoryStore
{
    private const int MaximumArchiveSamples = 100_000;
    private static readonly JsonSerializerOptions Json = new()
    {
        WriteIndented = false
    };

    private readonly string _path;
    private readonly string _lockPath;

    public QuotaHistoryStore(string? path = null)
    {
        _path = path ?? Path.Combine(
            AppSettings.SettingsDirectory,
            "weekly-quota-history-v1.json");
        _lockPath = Path.ChangeExtension(_path, ".lock");
    }

    public QuotaHistoryLoadResult Load(
        DateTimeOffset? endingAt = null)
    {
        try
        {
            using var gate = AcquireLock();
            return LoadUnlocked(endingAt ?? DateTimeOffset.Now);
        }
        catch
        {
            return new QuotaHistoryLoadResult(
                QuotaHistoryLoadStatus.Unavailable,
                new QuotaHistoryTimeline());
        }
    }

    public QuotaHistoryRecordResult Record(
        QuotaHistorySample sample,
        DateTimeOffset? endingAt = null)
    {
        var now = endingAt ?? DateTimeOffset.Now;
        using var gate = AcquireLock();
        var loaded = LoadUnlocked(now);
        var timeline = loaded.Status switch
        {
            QuotaHistoryLoadStatus.Absent =>
                new QuotaHistoryTimeline(),
            QuotaHistoryLoadStatus.Loaded => loaded.Timeline,
            QuotaHistoryLoadStatus.Corrupt =>
                throw new InvalidDataException(
                    "Quota history archive is corrupt and was not overwritten."),
            _ => throw new IOException(
                "Quota history archive is unavailable.")
        };

        var changed = timeline.Record(sample, now);
        if (changed) SaveUnlocked(timeline);
        return new QuotaHistoryRecordResult(timeline, changed);
    }

    private QuotaHistoryLoadResult LoadUnlocked(
        DateTimeOffset endingAt)
    {
        if (!File.Exists(_path))
            return new QuotaHistoryLoadResult(
                QuotaHistoryLoadStatus.Absent,
                new QuotaHistoryTimeline());

        try
        {
            var bytes = File.ReadAllBytes(_path);
            var archive = JsonSerializer.Deserialize<Archive>(bytes);
            if (archive is null
                || !IsArchiveValid(
                    archive.Version,
                    archive.Samples))
                return new QuotaHistoryLoadResult(
                    QuotaHistoryLoadStatus.Corrupt,
                    new QuotaHistoryTimeline());

            var timeline =
                new QuotaHistoryTimeline(archive.Samples);
            timeline.Prune(endingAt);
            return new QuotaHistoryLoadResult(
                QuotaHistoryLoadStatus.Loaded,
                timeline);
        }
        catch (JsonException)
        {
            return new QuotaHistoryLoadResult(
                QuotaHistoryLoadStatus.Corrupt,
                new QuotaHistoryTimeline());
        }
        catch (NotSupportedException)
        {
            return new QuotaHistoryLoadResult(
                QuotaHistoryLoadStatus.Corrupt,
                new QuotaHistoryTimeline());
        }
    }

    internal static bool IsArchiveValid(
        int version,
        IReadOnlyList<QuotaHistorySample>? samples) =>
        version == QuotaHistoryTimeline.ArchiveVersion
        && samples is not null
        && samples.Count <= MaximumArchiveSamples
        && samples.All(QuotaHistoryTimeline.IsValid)
        && samples.Zip(
                samples.Skip(1),
                (left, right) =>
                    left.Timestamp < right.Timestamp)
            .All(valid => valid);

    private FileStream AcquireLock()
    {
        var directory = Path.GetDirectoryName(_path);
        if (string.IsNullOrWhiteSpace(directory))
            throw new IOException(
                "Quota history directory is unavailable.");
        PrivateStorageSecurity.EnsureDirectory(directory);

        IOException? lastFailure = null;
        for (var attempt = 0; attempt < 80; attempt++)
        {
            try
            {
                var stream = new FileStream(
                    _lockPath,
                    FileMode.OpenOrCreate,
                    FileAccess.ReadWrite,
                    FileShare.None,
                    1,
                    FileOptions.WriteThrough);
                PrivateStorageSecurity.EnsureFile(_lockPath);
                return stream;
            }
            catch (IOException ex)
            {
                lastFailure = ex;
                Thread.Sleep(25);
            }
        }
        throw new IOException(
            "Quota history is being updated by another process.",
            lastFailure);
    }

    private void SaveUnlocked(QuotaHistoryTimeline timeline)
    {
        var archive = new Archive(
            QuotaHistoryTimeline.ArchiveVersion,
            timeline.Samples.ToList());
        var bytes = JsonSerializer.SerializeToUtf8Bytes(
            archive,
            Json);
        var temporary = _path
                        + $".{Environment.ProcessId}."
                        + $"{Guid.NewGuid():N}.tmp";
        try
        {
            using (var stream = new FileStream(
                       temporary,
                       FileMode.CreateNew,
                       FileAccess.Write,
                       FileShare.None,
                       4096,
                       FileOptions.WriteThrough))
            {
                stream.Write(bytes);
                stream.Flush(flushToDisk: true);
            }
            PrivateStorageSecurity.EnsureFile(temporary);
            if (File.Exists(_path))
                File.Replace(
                    temporary,
                    _path,
                    destinationBackupFileName: null,
                    ignoreMetadataErrors: true);
            else
                File.Move(temporary, _path);
            PrivateStorageSecurity.EnsureFile(_path);
        }
        finally
        {
            try
            {
                if (File.Exists(temporary))
                    File.Delete(temporary);
            }
            catch
            {
            }
        }
    }

    private sealed record Archive(
        int Version,
        List<QuotaHistorySample> Samples);
}
